src/app-logic/web-channel.js

```diff
@@ -3,54 +3,96 @@
+  | QuerySymbolicationApiMessageToBrowserBody;
+
+type StatusQueryMessageToBrowserBody = {| type: 'STATUS_QUERY' |};
+type EnableMenuButtonMessageToBrowserBody = {| type: 'ENABLE_MENU_BUTTON' |};
+type GetProfileMessageToBrowserBody = {| type: 'GET_PROFILE' |};
+type GetSymbolTableMessageToBrowserBody = {|
+  type: 'GET_SYMBOL_TABLE',
+  debugName: string,
+  breakpadId: string,
+|};
+type QuerySymbolicationApiMessageToBrowserBody = {|
+  type: 'QUERY_SYMBOLICATION_API',
+  path: string,
+  requestJson: string,
 |};
 
-/**
- * The messages are typed as an object so that the "type" field can be extracted
- * using the $Keys utility type.
- */
-type MessageFromBrowserObject = {|
-  STATUS_RESPONSE: {|
-    type: 'STATUS_RESPONSE',
-    menuButtonIsEnabled: boolean,
-    requestId: number,
-  |},
-  ENABLE_MENU_BUTTON_DONE: {|
-    type: 'ENABLE_MENU_BUTTON_DONE',
-    requestId: number,
-  |},
+export type MessageFromBrowser<R: ResponseFromBrowser> =
+  | OutOfBandErrorMessageFromBrowser
+  | ErrorResponseMessageFromBrowser
+  | SuccessResponseMessageFromBrowser<R>;
+
+type OutOfBandErrorMessageFromBrowser = {|
+  errno: number,
+  error: string,
+|};
+
+type ErrorResponseMessageFromBrowser = {|
+  type: 'ERROR_RESPONSE',
+  requestId: number,
+  error: string,
 |};
 
-// Extract out the different values. Exported for tests.
-export type MessageToBrowser = $Values<MessageToBrowserObject>;
-export type MessageFromBrowser = $Values<MessageFromBrowserObject>;
-export type MessageFromBrowserTypes = $Keys<MessageFromBrowserObject>;
+type SuccessResponseMessageFromBrowser<R: ResponseFromBrowser> = {
+  type: 'SUCCESS_RESPONSE',
+  requestId: number,
+  response: R,
+};
+
+export type ResponseFromBrowser =
+  | ResponseToStatusQuery
+  | ResponseToEnableMenuButton
+  | ResponseToGetProfile
+  | ResponseToGetSymbolTable
+  | ResponseToQuerySymbolicationApi;
+
+type ResponseToStatusQuery = {|
+  menuButtonIsEnabled: boolean,
+  // The version indicates which message types are supported by the browser.
+  // No version:
+  //   Shipped in Firefox 76.
+  //   Supports the following message types:
+  //    - STATUS_QUERY
+  //    - ENABLE_MENU_BUTTON
+  // Version 1:
+  //   Shipped in Firefox 93.
+  //   Adds support for the following message types:
+  //    - GET_PROFILE
+  //    - GET_SYMBOL_TABLE
+  //    - QUERY_SYMBOLICATION_API
+  version?: number,
+|};
+type ResponseToEnableMenuButton = void;
+type ResponseToGetProfile = ArrayBuffer | MixedObject;
+type ResponseToGetSymbolTable = SymbolTableAsTuple;
+type ResponseToQuerySymbolicationApi = string;
 
 /**
  * Ask the browser if the menu button is enabled.
  */
 export async function queryIsMenuButtonEnabled(): Promise<boolean> {
-  type ExpectedResponse = $PropertyType<
-    MessageFromBrowserObject,
-    'STATUS_RESPONSE'
-  >;
-
-  const response: ExpectedResponse = await _sendMessageWithResponse({
+  const response = await _sendMessageWithResponse<_, ResponseToStatusQuery>({
```

Comment:
When I've seen the use of this `_`, I first wasn't sure. But Flow seems to be happy enough with that, and I checked that it was still checking that we wouldn't use a random thing...

But if you follow my previous suggestion you will only need the request type :-)